src/installer-generator/installer-generator.c

```diff
@@ -0,0 +1,162 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "blockdev-util.h"
+#include "device-util.h"
+#include "dropin.h"
+#include "generator.h"
+#include "initrd-util.h"
+#include "parse-util.h"
+#include "path-util.h"
+#include "proc-cmdline.h"
+#include "special.h"
+#include "volatile-util.h"
+
+static const char *arg_dest = NULL;
+static enum {
+        INSTALLER_NO,
+        INSTALLER_YES,
+        INSTALLER_AUTO,
+} arg_enabled = INSTALLER_NO;
+static VolatileMode arg_volatile = VOLATILE_YES;
+
+static int parse_cmdline(const char *key, const char *value, void *data) {
+        int r;
+
+        assert(key);
+
+        if (proc_cmdline_key_streq(key, "systemd.installer")) {
+                if (!value)
+                        arg_enabled = INSTALLER_YES;
+                else if (streq(value, "auto"))
+                        arg_enabled = INSTALLER_AUTO;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer value \"%s\", ignoring: %m", value);
+                        arg_enabled = r ? INSTALLER_YES : INSTALLER_NO;
+                }
+        } else if (proc_cmdline_key_streq(key, "systemd.installer_volatile")) {
+                if (!value)
+                        arg_volatile = VOLATILE_YES;
+                else if (streq(value, "overlay"))
+                        arg_volatile = VOLATILE_OVERLAY;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer_volatile value \"%s\", ignoring: %m", value);
+                        else
+                                arg_volatile = r ? VOLATILE_YES : VOLATILE_NO;
+                }
+        }
+
+        return 0;
+}
+
+static int reconfigure_repart(void) {
+        return write_drop_in_format(arg_dest, "systemd-repart.service", 50, "installer",
+                                    "# Automatically generated by systemd-installer-generator\n\n"
+                                    "[Service]\n"
+                                    "ExecStart=\n"
+                                    "ExecStart=" ROOTBINDIR "/systemd-repart --dry-run=no --installer=%s\n",
```

Comment:
I tried that approach but it has a couple problems:

- It's impossible to support the `auto` mode that way
- It breaks the /sysusr or /sysroot logic, and it's impossible to re-implement that from the generator

Reply:
> I tried that approach but it has a couple problems:
>
> * It's impossible to support the auto mode that way

Can't this be checked directly within the generator (like it is already done for the target activation)? If it detects removable media you use the alternative definitions, if not, you simply omit the argument.

However, I was wondering what would be a use case where having separate repart.d files during installation is necessary?
Currently none come to my mind which make sense or cannot already be solved in other ways:

1. More partitions during installations than during runtime: No reason to do this as having the .d files also present during runtime would be a no-op.
2. Omit partitions from installation and only set them up during runtime: Upon reboot it would be the same as if they had been included during installation. Would only shift some time from installation to runtime in cases where setting up partition takes a while. (There is a small difference during the first time the initrd is invoked and thus before systemd-repartd.service might run but putting logic there seems like a bad idea anyhow.)
3. Different partition definitions during installation and runtime: Makes little sense - why should runtime have different definitions if repartd does not modify existing partitions anyhow?

Furthermore if someone really wants to have different different partitions in some scenario which did not came to my mind they would still be able to copy the corresponding files into place after installation.

Now there is another aspect which is talked mentioned in the docs but I think that can also be already solved with just the installer target: Suppressing partitioning of the removable media itself. If the generator detects removable media it can mask the unit or set dry-run to yes. This would suppress any partitioning. Now there might be cases distribution maintainers would like to have some stuff partitioned in which case they could add a new service pulled in by the installation target. Then they could the files somewhere else and partitioning of the installation media would still be a deliberate, opt-in thing.